WordPress/Classes/Utility/BackgroundTasks/WeeklyRoundupBackgroundTask.swift

```diff
@@ -34,86 +37,103 @@ private class WeeklyRoundupDataProvider {
 
-            statsServiceRemote.getData(for: .week, endingOn: periodEndDate, limit: 1) { (timeStats: StatsSummaryTimeIntervalData?, error) in
+            self.fetchStats(for: site, endingOn: periodEndDate, with: service) { [weak self] result in
                 defer {
-                    statsProcessed = statsProcessed + 1
-
-                    if statsProcessed == sites.count {
-                        let bestBlogStats = self.filterBest(5, from: blogStats)
-
-                        completion(.success(bestBlogStats))
-                    }
+                    group.leave()
                 }
-
-                guard let timeStats = timeStats else {
-                    guard let error = error else {
-                        self.onError(DataRequestError.unknownErrorRetrievingStats(site))
-                        return
-                    }
-
-                    self.onError(DataRequestError.errorRetrievingStats(dotComID, error: error))
+                guard let self else {
                     return
                 }
-
-                guard let stats = timeStats.summaryData.first else {
-                    // No stats for this site, or not enough views to qualify.  This is not an error.
-                    return
+                switch result {
+                case .failure(let error):
+                    self.onError(error)
+                case.success(let stats):
+                    if let stats {
+                        blogStats[site.managedObjectID] = stats
+                    }
                 }
+            }
+        }
+
+        group.notify(queue: .main) {
+            let bestBlogStats = self.filterBest(5, from: blogStats)
+            completion(.success(bestBlogStats))
+        }
+    }
 
-                blogStats[site] = stats
+    /// Fetches weekly stats data for a given site.
+    ///
+    /// This function fetches the stats for a single site and passes the result to a completion handler.
+    /// If it encounters any error during fetching, it calls the completion handler with an appropriate error object.
+    /// The completion handler is executed in the same queue as the provided `NSManagedObjectContext`.
+    private func fetchStats(
+        for site: Site,
+        endingOn periodEndDate: Date,
+        with service: StatsServiceRemoteV2,
+        completion: @escaping (Result<StatsSummaryData?, Error>) -> Void
+    ) {
+        service.getData(for: .week, endingOn: periodEndDate, limit: 1) { (timeStats: StatsSummaryTimeIntervalData?, error) in
+            guard let timeStats = timeStats else {
+                if let error = error {
+                    completion(.failure(DataRequestError.errorRetrievingStats(site.dotComID, error: error)))
+                } else {
+                    completion(.failure(DataRequestError.unknownErrorRetrievingStats(site.managedObjectID)))
+                }
+                return
             }
+            completion(.success(timeStats.summaryData.first))
```

Comment:
Nitpick: It's probably just me, but I find nested guard and if is hard to follow. On top of that, we need to make sure the `completion` must be called once and once only. What do you think we refactor it a little bit into something like this?

```swift
let result: Result<StatsSummaryData?, Error>
if let error {
  result = .failure(.errorRetrieving..)
} else if let timeStats {
  result = .success(...)
} else {
  result = .failure(unknownError...)
}
completion(result)
```